### ISSUE TITLE
wildcard_enum_match_arm fail with non-exhaustive enums and OR patterns

### ISSUE THREAD

**Issue by repi** [rust-lang/rust-clippy#6862]
Lint name: [`wildcard_enum_match_arm`](https://rust-lang.github.io/rust-clippy/master/index.html#wildcard_enum_match_arm)

I tried this code ([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=43848454d08645d5af43bfd552300a03)):

```rust
#![warn(clippy::wildcard_enum_match_arm)]

fn main() {
    use std::io::ErrorKind; // this enum uses `[non_exhaustive]`
    let e = ErrorKind::Interrupted;
    
    match e {
        ErrorKind::NotFound => println!("oh NotFound"),
        ErrorKind::PermissionDenied => println!("oh PermissionDenied"),
        ErrorKind::ConnectionRefused => println!("oh ConnectionRefused"),
        ErrorKind::ConnectionReset => println!("oh ConnectionReset"),
        ErrorKind::ConnectionAborted => println!("oh ConnectionAborted"),
        ErrorKind::NotConnected => println!("oh NotConnected"),
        ErrorKind::AddrInUse => println!("oh AddrInUse"),
        ErrorKind::AddrNotAvailable => println!("oh AddrNotAvailable"),
        ErrorKind::BrokenPipe => println!("oh BrokenPipe"),
        ErrorKind::AlreadyExists => println!("oh AlreadyExists"),
        ErrorKind::WouldBlock => println!("oh WouldBlock"),
        ErrorKind::InvalidInput => println!("oh InvalidInput"),
        ErrorKind::InvalidData => println!("oh InvalidData"),
        ErrorKind::TimedOut => println!("oh TimedOut"),
        ErrorKind::WriteZero => println!("oh WriteZero"),
        ErrorKind::UnexpectedEof => println!("oh UnexpectedEof"),
        
        // works
        //ErrorKind::Interrupted => println!("some stuffs"),
        //ErrorKind::Other => println!("some stuffs"),

        // doesn't work
        ErrorKind::Interrupted |
        ErrorKind::Other => println!("some stuffs"),

        _ => println!("something else"),
    }
}
```

I expected to see this happen: The lint not triggering, as all enum variants including a non-exhaustive case are covered

Instead, this happened: The lint triggers and wants the `_` case changed to `std::io::ErrorKind::Interrupted | std::io::ErrorKind::Other | _`, but those concrete variant types are already handled.

### Meta

- `cargo clippy -V`: `clippy 0.0.212 (cb75ad5d 2021-02-10)`
- `rustc -Vv`:
```
rustc 1.50.0 (cb75ad5db 2021-02-10)
binary: rustc
commit-hash: cb75ad5db02783e8b0222fee363c5f63f7e2cf5b
commit-date: 2021-02-10
host: x86_64-pc-windows-msvc
release: 1.50.0
```
